pkg/operator/certrotation/signer_test.go

```diff
@@ -1,24 +1,219 @@
+func TestEnsureSigningCertKeyPairRace(t *testing.T) {
+	secondControllerStartChan := make(chan struct{})
+	secondControllerEndChan := make(chan struct{})
+	firstControllerEndChan := make(chan struct{})
+
+	// represents a secret that was created before 4.7 and
+	// hasn't been updated until now (upgrade to 4.15)
+	existingSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "ns", Name: "signer", ResourceVersion: "10",
+			Annotations: map[string]string{
+				"auth.openshift.io/certificate-not-after":  "2108-09-08T22:47:31-07:00",
+				"auth.openshift.io/certificate-not-before": "2108-09-08T20:47:31-07:00",
+			},
+		},
+		Type: "SecretTypeTLS",
+		Data: map[string][]byte{"tls.crt": {}, "tls.key": {}},
+	}
+	// add a key pair so that the check
+	// (GetCAFromBytes) in the
+	// target's method doesn't fail
+	if err := setSigningCertKeyPairSecret(existingSecret, 24*time.Hour); err != nil {
+		t.Fatal(err)
+	}
+
+	indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+	if err := indexer.Add(existingSecret); err != nil {
+		t.Fatal(err)
+	}
+
+	fakeKubeClient := kubefake.NewSimpleClientset(existingSecret)
+	fakeKubeClient.PrependReactor("delete", "secrets", func(action clienttesting.Action) (handled bool, ret runtime.Object, err error) {
+		// the assumption here is that delete operation
+		// will be delivered via watch to the second controller
+		// which could result in a 404 thus
+		// to simulate this scenario
+		// the secret si removed from the cache
+		if err = indexer.Delete(existingSecret); err != nil {
+			return true, nil, err
+		}
+		close(secondControllerStartChan) // this triggers the second controller
+		return false /*means that the next reactor in the chain will be invoked (secret will be deleted)*/, nil, nil
+	})
+	fakeKubeClient.PrependReactor("get", "secrets", func() clienttesting.ReactionFunc {
+		counter := 0
+		counterLock := sync.Mutex{}
+		return func(action clienttesting.Action) (bool, runtime.Object, error) {
+			counterLock.Lock()
+			defer counterLock.Unlock()
+			defer func() {
+				counter++
+			}()
+			if counter == 0 {
+				// this is the call from the first controller
+				return true, existingSecret, nil
+			}
+			// this is a call from the second controller
+			// wait until the first controller finishes
+			// so that we can update the secret with the new keys
+			<-firstControllerEndChan
+
+			// cannot use fakeKubeClient.ActionsA()
+			// because it deadlocks, fakeKubeClient.RLock already acquired
+			//
+			// the created secret from the actions
+			// is validated at the end of this test
+			// and must match this one
+			createdSecret := existingSecret.DeepCopy()
+			createdSecret.Type = corev1.SecretTypeTLS
+			createdSecret.Annotations["openshift.io/owning-component"] = "test"
```

Comment:
I don't understand this part. Won't the default behavior of the test clientset return the object you expect (i.e. the one created by the first controller) without this?